0023-communication-protocol.adoc

```diff
@@ -0,0 +1,388 @@
+We assume that the two parties have already established the roles of (cryptographic) Alice and Bob and those roles are an input to the communication protocol.
+
+To design a better communication protocol than the first one, it is useful to take a step back and think protocols in a more abstract manner.
+
+=== Protocols in COMIT
+
+A protocol can be roughly defined as the order and semantics of actions.
+
+Within COMIT, we have two kinds of protocols:
+
+. Cryptographic protocols
+. Communication protocols
+
+A cryptographic protocol within COMIT defines the order and semantics of interactions with blockchains.
+At the moment, COMIT specializes in cryptographic protocols that facilitate atomic swaps.
+
+With the https://github.com/comit-network/spikes/blob/master/0021-protocol-naming.adoc[protocol naming spike], we've made it more explicit that COMIT does not support one but several cryptographic protocols at the moment.
+In particular, we currently support:
+
+* HAN for Bitcoin
+* HAN for Ethereum
+* HErc20 for Ethereum
+
+We've managed to identify a protocol (HAN) that can be applied to both of the ledgers that we currently support.
+Even though this allows for some reuse, when you take a look at the actual interactions with the blockchain, HAN for Bitcoin and HAN for Ethereum differ (you cannot broadcast a Bitcoin transaction on Ethereum and vice versa).
+Given that there is always two protocols that need to be combined for an atomic swap, our total number of supported protocols (considering the full swap as a black box) is 6 (2 * 3).footnote:[We don't allow HAN(Ethereum)-HErc20 at the moment, so this number drops to 4 in reality.]
+
+A communication protocol within COMIT defines how two COMIT nodes interact with each other.
+Those (libp2p-based) protocols facilitate the exchange of information for the purpose of then following a cryptographic protocol.
+
+These two kinds of protocols have a relationship to each other.
+In the past we went with a N-to-1 relationship: we designed a single communication protocol that was supposed to handle communication for all kinds of cryptographic protocols.
+Even though we sweated blood in the design of the original communication protocol, we quickly learned that it is not as extensible and flexible as we thought it would to be.
+
+=== What is the ideal relationship between cryptographic and libp2p protocols?
+
+The main problem with the design of the original communication protocol was that we tried to anticipate the future a bit too much.
+Even thought it is important to allow for future extensions, one has to be careful about how the extension mechanism works.
+
+Let's consider the following requirement:
+
+. In order to execute a certain cryptographic protocol, COMIT nodes may want to exchange information.
```

Comment:
I am not sure I understand your point here.

Reply:
Let me try to add some information, let me know if it helps:

Cryptographic protocols may require information exchange such as e.g. the secret-hash and certain identities for HAN. But the information exchanged may differ between different cryptographic protocols. 

Thus,  the question is, how to design the **communication protocol** given different information exchange in cryptographic protocols? Thomas talks about 2 extremes, we can create the communication protocol very generic so one can pack different information into one message of one protocol, or we can split it into several protocols that have a specific concern and only one message.

Reply:
I think my issue is more with the wording: "Let's consider the following requirement" invite to consider an example to make the explanation clearer to the reader.

Upon this section, I am more confused then if the example were not here.

Reply:
What this statement aimed to make clearer is that the communication protocol is _motivated_ by us wanting to execute cryptographic protocols. Maybe this is so obvious that it doesn't need mentioning?

Reply:
I've revised this section.